server/Controllers/GameControllers/snake.js

```diff
@@ -1,13 +1,126 @@
 class Snake {
+  /**
+   * @param  {string} playerId
+   * @param  {number} initialPositionX
+   * @param  {number} initialPositionY
+   * @param  {string} initialDirection
+   * @param  {string} snakeColor
+   * @param  {number} initialSnakeLength
+   * @param  {Function} eatFood Function to check that snake's next movement will be eating food.
+   * @param  {Function} hasCollidedWithRival Function to check that snake's next movement is colliding with Rival's body or not
+   * @param  {number} intervalNumber
+   */
   constructor(playerId, initialPositionX, initialPositionY,
-    initialDirection, snakeColor, initialSnakeLength) {
+    initialDirection, snakeColor, initialSnakeLength, eatFood, hasCollidedWithRival, intervalNumber) {
     this.ownerId = playerId;
     this.head = { x: initialPositionX, y: initialPositionY };
     this.direction = initialDirection;
     this.color = snakeColor;
     this.bodyLength = initialSnakeLength;
+    this.eatFood = eatFood;
+    this.hasCollidedWithRival = hasCollidedWithRival;
+    this.interval = intervalNumber; // interval variable, if snake collides we will clear this interval
+    this.xAxisVelocity = 0; // the x axis movement of snake in each interval
+    this.yAxisVelocity = 0; // the y axis movement of snake in each interval
     this.bodyCoordinates = [];
   }
+
+  /**
+   * This function generates snake body coordinates from the given starting head coordinates
+   */
+  createInitialSnakeBody() {
+    switch (this.direction) {
+      case 'down': {
+        this.yAxisVelocity = 2;
+        break;
+      }
+      case 'up': {
+        this.yAxisVelocity = -2;
+        break;
+      }
+      case 'left': {
+        this.xAxisVelocity = -2;
+        break;
+      }
+      case 'right': {
+        this.xAxisVelocity = 2;
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+    // FirstRender of snake body
+    for (let x = 0; x < this.length; x += 1) {
+      this.moveSnakeOneStep();
+    }
+  }
+
+  /**
+   * Moves the snakes one steps according to relative change on x axis or y axis.
+   */
+  moveSnakeOneStep() {
+    this.head.x += this.xAxisVelocity; // snake head's new x coordinate
+    this.head.y += this.yAxisVelocity; // snake head's new y coordinate
+    if (this.eatFood(this.head)) {
+      this.length += 15;
+      // @TODO implement the score/lifeline increment here.
+    }
+    if (this.hasCollidedWithItself()
+      || this.hasCollidedWithRival(this.head)) {
+      // @TODO emit a game mechanics event to show the modal for game lost.
+
+    } else {
+      this.pushCoordinates();
```

Comment:
this method is for growing the snake body right. It would be better if we called it `this.growSnake` or something

Reply:
but growSnake, doesn't signify how much snake body, we are increasing.

Reply:
moveSnakeOneStep sounds more coherent according to me?

Reply:
I was talking about `this.pushCoordinates` method. What does it do?

Reply:
the information about this function you can find on the function definition.

Reply:
I read it, I still don't understand. Why are you pushing new coordinates? Is it for growing the snake or some other reason?

Reply:
It is putting new coordinate in the snake body parameter and removing the tail coordinate, But I understand I will refactor the name and doc of the function.